### PR TITLE
Add missing-dependency & circular dependency error ctx

### PR DESCRIPTION
**Pull request by KGOH**
Example
```
ExceptionInfo: Missing dependency
  :stats-enabled
  {:type   :darkleaf.di.core/missing-dependency
   :parent foo.bar.core/transform
   :key    :stats-enabled}
```

### CODE REVIEW

File: src/darkleaf/di/core.clj [hunk 1]
```diff
@@ -86,7 +87,7 @@
   (let [ctx           (update ctx :under-construction conj key)
         factory       (registry key)
         declared-deps (p/dependencies factory)
-        resolved-deps (resolve-deps ctx declared-deps)
+        resolved-deps (resolve-deps (assoc ctx :*current-key key) declared-deps)
```
Comment (darkleaf) [inline src/darkleaf/di/core.clj:90]: `*` is a mark for volatile

File: src/darkleaf/di/core.clj [hunk 2]
```diff
@@ -55,10 +55,11 @@
...
 (declare find-or-build)
 
-(defn- missing-dependency! [key]
+(defn- missing-dependency! [{:as ctx :keys [*current-key]} key]
   (throw (ex-info (str "Missing dependency " key)
-                  {:type ::missing-dependency
-                   :key  key})))
+                  {:type   ::missing-dependency
+                   :key    key
+                   :parent *current-key})))
```
Comment (darkleaf) [inline src/darkleaf/di/core.clj:62]: may be you could just use `:under-construction` ?

File: src/darkleaf/di/core.clj [hunk 3]
```diff
@@ -55,10 +55,11 @@
 
 (declare find-or-build)
 
-(defn- missing-dependency! [key]
+(defn- missing-dependency! [{:as ctx :keys [*current-key]} key]
```
Comment (darkleaf) [inline src/darkleaf/di/core.clj:58]: could you please add tests?

File: src/darkleaf/di/core.clj (hunk 2 above)
Comment (KGOH) [reply src/darkleaf/di/core.clj:62]: @darkleaf :under-construction is a set, you can't get last added element from it

File: src/darkleaf/di/core.clj (hunk 1 above)
Comment (KGOH) [reply src/darkleaf/di/core.clj:90]: @darkleaf I can remove `*`, I just used already existing key:
 https://github.com/darkleaf/di/blob/497a2497dc171c0556f3f48df355090decfcfbbd/src/darkleaf/di/core.clj#L85C45-L85C57

**Review by darkleaf — APPROVED** [review]

**Pull request merged**